Plugins/Public/docking_modules/Utilities.cpp

```diff
@@ -0,0 +1,306 @@
+void ReadFLFile(map<string, vector<string>> &variables, string &path)
+{
+	ifstream file(path);
+	if (file.is_open())
+	{
+		string line;
+		while (getline(file, line))
+		{
+			for (map<string, vector<string>>::iterator it = variables.begin(); it != variables.end(); ++it)
+			{
+				if (boost::algorithm::starts_with(line, it->first))
+				{
+					uint pos = it->first.size();
+
+					// Reject partial matches.
+					if (line[pos] == ' ')
+					{
+						// pos + 3 - we skip whitespace, equation symbol and whitespace.
+						it->second.push_back(line.substr(pos + 3));
+						break;
+					}
+				}
+			}
+		}
+
+		file.close();
+	}
+}
+
+// Expanded version of previous function.
+void ReadFLFile(map<string, vector<string>> &variables, map<string, string> &hookExtData, string &path)
+{
+	ifstream file(path);
+	if (file.is_open())
+	{
+		string line;
+		while (getline(file, line))
+		{
+			if (boost::algorithm::starts_with(line, "[flhook]"))
+				goto HookExtCheck;
+
+			for (map<string, vector<string>>::iterator it = variables.begin(); it != variables.end(); ++it)
+			{
+				if (boost::algorithm::starts_with(line, it->first))
+				{
+					uint pos = it->first.size();
+
+					// Reject partial matches.
+					if (line[pos] == ' ')
+					{
+						// pos + 3 - we skip whitespace, equation symbol and whitespace.
+						it->second.push_back(line.substr(pos + 3));
+						break;
+					}
+				}
+			}
+		}
+
+		// No lines to read remain, avoid HookExtCheck and go directly to exit procedure.
+		goto Exit;
+
+	HookExtCheck:
+		while (getline(file, line))
+		{
+			uint pos = line.find(" ");
+			hookExtData.insert(make_pair(line.substr(0, pos), line.substr(pos + 3)));
+		}
+
+	Exit:
+		file.close();
+	}
+}
+
+// Converts text in UCS-2 FL encoding to hex string.
+string DecodeWStringToStringOfBytes(wstring &wstr)
+{
+	string bytestr;
+	bytestr.reserve(wstr.size() * 4);
+
+	for (wchar_t c : wstr)
+	{
+		char buf[4];
+		sprintf(buf, "%04x", c);
+		bytestr += buf;
+	}
+
+	return bytestr;
+}
+
+// Converts hex string to UCS-2 text back.
+wstring EncodeWStringFromStringOfBytes(string &bytestr)
+{
+	wstring charName;
+
+	for (uint i = 0; i < bytestr.size() / 4; ++i)
+	{
+		charName.push_back((wchar_t)strtol(bytestr.substr(i * 4, 4).c_str(), NULL, 16));
+	}
+
+	return charName;
+}
+
+// Get vector of strings from one string by separating it by some char.
+vector<string> GetParams(string &str, char splitChar)
+{
+	vector<string> Params;
+	uint lastPos = 0;
+
+	if (str.empty())
+		return Params;
+
+	for (uint i = 1; true; ++i)
+	{
+		if (str[i] == splitChar)
+		{
+			Params.push_back(str.substr(lastPos, i - lastPos));
+			lastPos = i + 1;
+		}
+
+		if (i == str.size() - 1)
+		{
+			Params.push_back(str.substr(lastPos));
+
+			if (Params.empty())
+				Params.push_back(str);
+
+			break;
+		}
+	}
+
+	return Params;
+}
+
+// Convert multiple strings to one by uniting them with some char.
+string SetParams(vector<string> &Params, char uniteChar)

```

Comment:
This looks like a simple implode/join function, surely Boost has something for this.

Reply:
Yes, boost variant seems good.